File: day1.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <numeric>

using namespace std;
using Vector = vector<int>;

int part1(Vector& numbers)
{
    int previous = 0;
    unsigned larger = 0;
    for (auto number : numbers) {
        if (previous && number > previous) {
            larger++;
        }
        previous = number;
    }
    return larger;
}

int part2(Vector& numbers)
{
    Vector sums;
    Vector::const_iterator begin, end;
    constexpr auto windowSize = 3;
    for (begin = numbers.begin(), end = begin + windowSize; end != numbers.end(); begin++, end++) {
        sums.push_back(accumulate(begin, end, 0));
    }
    return part1(sums);
}

int main(int argc, char* argv[])
{
    ifstream input;
    input.open(argc < 2 ? "day1.txt" : argv[1], ifstream::in);

    Vector numbers;
    int number;
    while (input.good()) {
        input >> number;
        numbers.push_back(number);
    }

    cout << "Part 1: " << part1(numbers) << "\n";
    cout << "Part 2: " << part2(numbers) << "\n";
}

File: day1.py
import sys
from collections import deque

def count_larger(scan):
    previous = None
    larger = 0
    for line in scan:
        current = int(line)
        if previous and current > previous:
            larger += 1
        previous = current
    return larger

def count_windowed(scan):
    sums = []
    window = deque()
    for line in scan:
        window.append(int(line))
        if len(window) > 3:
            window.popleft()
        if len(window) == 3:
            sums.append(sum(window))
    return count_larger(sums)

def main(args = ()):
    arg = "day1.txt" if len(args) < 1 else args[0]
    with open(arg, "r") as scan:
        print("Part 1:", count_larger(scan))
        scan.seek(0)
        print("Part 2:", count_windowed(scan))


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))